### ISSUE TITLE
EF TPH scenario: AssertConfigurationIsValid() does not evaluate AfterMap()

### ISSUE THREAD

**Issue by timmkrause** [AutoMapper/AutoMapper#2084]
### Source/destination types

```csharp
// Source, part of DbSet in DbContext, returned by repository.
public abstract class Decision
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public Guid Id { get; set; }

    // ...
}

// Source (actually used at runtime). When using TPH (table per hierarchy) with Entity Framework this is the actual type at runtime.
public class OtherDecision : Decision
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public Guid Id { get; set; }

    public string Reason { get; set; }

    // ...
}

// Target
public class DecisionContract
{
    public Guid Id { get; set; }

    public string Reason { get; set; }
}
```

### Mapping configuration

```csharp
public class DecisionContractProfile : Profile
{
    public DecisionContractProfile()
    {
        this.CreateMap<Decision, DecisionContract>()

            // To make AssertConfigurationIsValid unit test work again, obsiously doesn't evaluate AfterMap().
            .ForMember(dest => dest.Reason, opt => opt.Ignore())
            .AfterMap(
                (src, dest) =>
                {
                    var otherDecision = src as OtherDecision;

                    if (otherDecision != null)
                    {
                        dest.Reason = otherDecision.Reason;
                    }
                });
    }
}
```

### Version: x.y.z
6.0.2

### Expected behavior
`AssertConfigurationIsValid()` should evaluate `AfterMap()` correctly and detect that target property is being mapped.

### Actual behavior
```
AutoMapper.AutoMapperConfigurationException : 
Unmapped members were found. Review the types and members below.
Add a custom mapping expression, ignore, add a custom resolver, or modify the source/destination type
For no matching constructor, add a no-arg ctor, add optional arguments, or map all of the constructor parameters
===========================================================================================
Decision -> DecisionContract (Destination member list)

Unmapped properties:
Reason
```

### Steps to reproduce
- Use TPH (table per hierarchy) with Entity Framework and map the abstract source type (which is the specialized type at runtime) to the destination type.
- Use `AfterMap()` to identify the actual type and map the missing property manually.
- `AssertConfigurationIsValid()` throws.

**Comment by timmkrause** [comment]
Btw.: I could totally understand if you don't want to support such scenarios, could potentially result in feature creep. I am totally fine with `opt.Ignore()` in this case. Just wanted to let you know.

**Comment by lbargaoanu** [comment]
AfterMap is just an event, totally opaque to AM. If you want validation, you need to use smth relevant for AM, a resolver for example.

**Comment by jbogard** [comment]
There are two main reasons why we don't evaluate custom functions during mapping configuration:

1. The mapping functions may need contextual information only available at mapping time. This means calling Map would throw an exception. For example, resolvers/type converters may use dependency injection to grab EF DbContexts.
2. We would have to somehow detect what's been set in all the different places you completely take over mapping

During configuration validation, we don't execute mappings. We only look at the TypeMap/PropertyMap configuration objects themselves.

The only way we can really "know" what you're mapping to/from is to use expression trees since we can inspect these at runtime. Since lambda expressions are the only thing that can form an Expression, and not lambda statements, we can't peer inside your AfterMap

**tl;dr what you're asking for is impossible.**

And as a side note, I would only use AfterMap if there is literally nothing else that can work for me, or I can only do something after the entire source/destination objects have been mapped. In your case, I'd do an Ignore() on the base mapping but also create a map in the derived case: `CreateMap<OtherDecision, DecisionContract>`

AutoMapper knows how to handle polymorphic mappings.
